### PR TITLE
Ignore incomplete trials in `matplotlib.plot_parallel_coordinate`

### PR DESCRIPTION
**Pull request by nzw0301**
## Motivation

- Followup of https://github.com/optuna/optuna/pull/3373
- Current implementation, introducing nan value for missing param, makes debugging/testing hard.

## Description of the changes

- Apply the same logic to `visualize.matplotlib.plot_parallel_coordinate`

---

Matplotlib's test should be addressed by #3368...

### CODE REVIEW

**Comment by nzw0301** [comment]
Oh, the output looks strange. The maximum objective value should be `1.0`.

```python
from optuna.testing.visualization import prepare_study_with_trials
import optuna

study = prepare_study_with_trials(with_c_d=False)

for t in study.get_trials():
    print(f"trial id:{t.number}, objective:{t.value}, params:{t.params}")

# the outputs:
# trial id:0, objective:0.0, params:{'param_a': 1.0, 'param_b': 2.0}
# trial id:1, objective:2.0, params:{'param_b': 0.0}
# trial id:2, objective:1.0, params:{'param_a': 2.5, 'param_b': 1.0}

# plotly for reference, 
optuna.visualization.plot_parallel_coordinate(study)

optuna.visualization.matplotlib.plot_parallel_coordinate(study)
```
download (1)

**Comment by nzw0301** [comment]
Oh, the output looks this:

```python
from optuna.testing.visualization import prepare_study_with_trials
import optuna

study = prepare_study_with_trials(with_c_d=False)

for t in study.get_trials():
    print(f"trial id:{t.number}, objective:{t.value}, params:{t.params}")

# the outputs:
# trial id:0, objective:0.0, params:{'param_a': 1.0, 'param_b': 2.0}
# trial id:1, objective:2.0, params:{'param_b': 0.0} <--- this trial is ignored because `param_b` is missing.
# trial id:2, objective:1.0, params:{'param_a': 2.5, 'param_b': 1.0}

# plotly for reference, 
optuna.visualization.plot_parallel_coordinate(study)

optuna.visualization.matplotlib.plot_parallel_coordinate(study)
```
download (1)

**Comment by hvy** [comment]
@himkt , @HideakiImamura could you review this PR? Please reassign if you're busy.

**Comment by HideakiImamura** [comment]
@nzw0301 I generated the plot using the code provided by you. The plot by this PR misses the trial#2 and includes the trial#1 (this one should not be included). The plot by the current master misses the trial#1 and includes trial#2. The behavior of the current master is expected. Am I misunderstanding something?

- code
```python
import matplotlib.pyplot as plt
from optuna.testing.visualization import prepare_study_with_trials
import optuna

study = prepare_study_with_trials(with_c_d=False)

for t in study.get_trials():
    print(f"trial id:{t.number}, objective:{t.value}, params:{t.params}")

# the outputs:
# trial id:0, objective:0.0, params:{'param_a': 1.0, 'param_b': 2.0}
# trial id:1, objective:2.0, params:{'param_b': 0.0} <--- this trial is ignored because `param_b` is missing.
# trial id:2, objective:1.0, params:{'param_a': 2.5, 'param_b': 1.0}

# plotly for reference, 
fig = optuna.visualization.plot_parallel_coordinate(study)
fig.write_image("./plotly.png")

fig = optuna.visualization.matplotlib.plot_parallel_coordinate(study)
plt.savefig("./plt.png")

```
- This PR
plt

- Current master
master_plt

**Comment by nzw0301** [comment]
Indeed, I should have explained this change with the current example.

Both plots do not show trial#1 as a line plot, but the master version uses trial#1 to determine the max/min value for `objective` and `param_b` vertical lines, as a result, the generated plot has empty space: specifically 
- objective: 1--2
- param_b: 0--1

On the other hand, this PR does not use trial#1's information to decide max/min values. I believe this version is more informative and more consistent with [the explanation of the documentation](https://optuna.readthedocs.io/en/v2.0.0/reference/generated/optuna.visualization.plot_parallel_coordinate.html)

> Note that, If a parameter contains missing values, a trial with missing values is not plotted.

than the current version 

As a more aggressive example, the following code adds a trial, trial#4, whose objective is `1000` and suggested parameters are nothing to the study in the example above: 

```python
import matplotlib.pyplot as plt
from optuna.testing.visualization import prepare_study_with_trials
import optuna
import seaborn as sns
sns.set_theme()

study = prepare_study_with_trials(with_c_d=False)

# add trial#4
def objective(t):
    return 1000.
study.optimize(objective, n_trials=1)

for t in study.get_trials():
    print(f"trial id:{t.number}, objective:{t.value}, params:{t.params}")
# the outputs:
# trial id:0, objective:0.0, params:{'param_a': 1.0, 'param_b': 2.0}
# trial id:1, objective:2.0, params:{'param_b': 0.0} <--- this trial is ignored because `param_b` is missing.
# trial id:2, objective:1.0, params:{'param_a': 2.5, 'param_b': 1.0}
# trial id:3, objective:1000.0, params:{} <--- this trial is ignored because `param_a` and`param_b` are missing.


# plotly for reference, 
fig = optuna.visualization.plot_parallel_coordinate(study)
fig.write_image("./plotly.png")

fig = optuna.visualization.matplotlib.plot_parallel_coordinate(study)
plt.savefig("./plt.png")
```
The results are as follows:

The master version uses trial#4, so the plot is not brilliant in my opinion.
plt

This branch: 
plt

Plotly version on master:
plotly

File: optuna/visualization/_utils.py [hunk 1]
```diff
@@ -113,6 +114,33 @@ def _get_param_values(trials: List[FrozenTrial], p_name: str) -> List[Any]:
     return list(map(str, values))
 
 
+def _get_skipped_trial_numbers(
+    trials: List[FrozenTrial], used_param_names: Sequence[str]
+) -> Set[int]:
+    """Utility function for ``plot_parallel_coordinate``.
+
+    If trial's parameters does not contain a parameter in ``used_param_names``,
```
Comment (HideakiImamura) [inline optuna/visualization/_utils.py:122]: A nit
```suggestion
    If trial's parameters do not contain a parameter in ``used_param_names``,
```

File: optuna/visualization/_utils.py [hunk 2]
```diff
@@ -113,6 +114,33 @@ def _get_param_values(trials: List[FrozenTrial], p_name: str) -> List[Any]:
...
+    ``plot_parallel_coordinate`` methods do not use such trails.
+
+    Args:
+        trials:
+            List of ``FrozenTrials``.
```
Comment (HideakiImamura) [inline optuna/visualization/_utils.py:127]:
```suggestion
            List of ``FrozenTrial``s.
```

**Review by HideakiImamura — COMMENTED** [review]
Thanks for the clear explanation. I misunderstood, but now I've understood. Have several nit comments. PTAL.

**Comment by nzw0301** [comment]
Thanks! I've updated the PR by accepting your all suggestions.

**Review by HideakiImamura — APPROVED** [review]
Thanks and sorry for the late. LGTM!

**Review by himkt — APPROVED** [review]
LGTM, thank you for the improvement.

As a side note, I'm wondering if we can unify these handlings (i.e. instead of retrieving a list of numbers of trials missing parameters, filtering trials (remaining trials whose parameters are not missed) and use it in subsequent procedures).

```python
    # Prepare data for plotting.
    trials = [trial for trial in study.trials if trial.state == TrialState.COMPLETE]

    all_params = {p_name for t in trials for p_name in t.params.keys()}
    if params is not None:
        for input_p_name in params:
            if input_p_name not in all_params:
                raise ValueError("Parameter {} does not exist in your study.".format(input_p_name))
        all_params = set(params)
    sorted_params = sorted(all_params)

    trials = filter_trials(trials, sorted_params)

    if len(trials) == 0:
        _logger.warning("Your study does not have any completed trials.")
        return ax

    obj_org = [target(t) for t in trials]
```

With the change, I think we can remove `if t.number not in skipped_trial_numbers` from the codebase. But the change is solid enough so let me merge. Sorry for blocking the works with the delayed response. :bow:

https://github.com/optuna/optuna/blob/4310ca7200b20013c1892baf927ebcf20ff83793/optuna/visualization/matplotlib/_parallel_coordinate.py#L122-L124

https://github.com/optuna/optuna/blob/4310ca7200b20013c1892baf927ebcf20ff83793/optuna/visualization/matplotlib/_parallel_coordinate.py#L138-L140

**Pull request merged**